\hypertarget{group__SHAPE__reactivities}{}\section{Incorporating S\+H\+A\+P\+E reactivity data}
\label{group__SHAPE__reactivities}\index{Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}}


Incorporate S\+H\+A\+P\+E reactivity structure probing data into the folding recursions by means of soft constraints.  


Collaboration diagram for Incorporating S\+H\+A\+P\+E reactivity data\+:
\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=350pt]{group__SHAPE__reactivities}
\end{center}
\end{figure}
\subsection*{Files}
\begin{DoxyCompactItemize}
\item 
file \hyperlink{constraints__SHAPE_8h}{constraints\+\_\+\+S\+H\+A\+P\+E.\+h}
\begin{DoxyCompactList}\small\item\em This module provides function to incorporate S\+H\+A\+P\+E reactivity data into the folding recursions by means of soft constraints. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Functions}
\begin{DoxyCompactItemize}
\item 
int \hyperlink{group__SHAPE__reactivities_ga57d612b58e1c61dd6cfcb5a843f8f1b3}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan} (\hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$vc, const double $\ast$reactivities, double m, double b, unsigned int options)
\begin{DoxyCompactList}\small\item\em Add S\+H\+A\+P\+E reactivity data as soft constraints (Deigan et al. method) \end{DoxyCompactList}\item 
int \hyperlink{group__SHAPE__reactivities_ga04ba85da63d8c793bb8001d1e6f800ba}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali} (\hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$vc, const char $\ast$$\ast$shape\+\_\+files, const int $\ast$shape\+\_\+file\+\_\+association, double m, double b, unsigned int options)
\begin{DoxyCompactList}\small\item\em Add S\+H\+A\+P\+E reactivity data from files as soft constraints for consensus structure prediction (Deigan et al. method) \end{DoxyCompactList}\item 
int \hyperlink{group__SHAPE__reactivities_gaf3c65a045060aef5c4e41693d30af58c}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam} (\hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$vc, const double $\ast$reactivities, double b, double default\+\_\+value, const char $\ast$shape\+\_\+conversion, unsigned int options)
\begin{DoxyCompactList}\small\item\em Add S\+H\+A\+P\+E reactivity data as soft constraints (Zarringhalam et al. method) \end{DoxyCompactList}\item 
int \hyperlink{group__SHAPE__reactivities_ga67675b3ed48744489a3bcfa4174197cb}{vrna\+\_\+sc\+\_\+\+S\+H\+A\+P\+E\+\_\+to\+\_\+pr} (const char $\ast$shape\+\_\+conversion, double $\ast$values, int length, double default\+\_\+value)
\begin{DoxyCompactList}\small\item\em Convert S\+H\+A\+P\+E reactivity values to probabilities for being unpaired. \end{DoxyCompactList}\end{DoxyCompactItemize}


\subsection{Detailed Description}
Incorporate S\+H\+A\+P\+E reactivity structure probing data into the folding recursions by means of soft constraints. 



\subsection{Function Documentation}
\hypertarget{group__SHAPE__reactivities_ga57d612b58e1c61dd6cfcb5a843f8f1b3}{}\index{Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}!vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan@{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan}}
\index{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan@{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan}!Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}}
\subsubsection[{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan}]{\setlength{\rightskip}{0pt plus 5cm}int vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan (
\begin{DoxyParamCaption}
\item[{{\bf vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$}]{vc, }
\item[{const double $\ast$}]{reactivities, }
\item[{double}]{m, }
\item[{double}]{b, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\label{group__SHAPE__reactivities_ga57d612b58e1c61dd6cfcb5a843f8f1b3}


{\ttfamily \#include $<$\hyperlink{constraints__SHAPE_8h}{Vienna\+R\+N\+A/constraints\+\_\+\+S\+H\+A\+P\+E.\+h}$>$}



Add S\+H\+A\+P\+E reactivity data as soft constraints (Deigan et al. method) 

This approach of S\+H\+A\+P\+E directed R\+N\+A folding uses the simple linear ansatz \[ \Delta G_{\text{SHAPE}}(i) = m \ln(\text{SHAPE reactivity}(i)+1)+ b \] to convert S\+H\+A\+P\+E reactivity values to pseudo energies whenever a nucleotide $ i $ contributes to a stacked pair. A positive slope $ m $ penalizes high reactivities in paired regions, while a negative intercept $ b $ results in a confirmatory ``bonus\textquotesingle{}\textquotesingle{} free energy for correctly predicted base pairs. Since the energy evaluation of a base pair stack involves two pairs, the pseudo energies are added for all four contributing nucleotides. Consequently, the energy term is applied twice for pairs inside a helix and only once for pairs adjacent to other structures. For all other loop types the energy model remains unchanged even when the experimental data highly disagrees with a certain motif.

\begin{DoxySeeAlso}{See also}
For further details, we refer to \cite{deigan:2009}. 

\hyperlink{group__soft__constraints_ga73cdc07b9a199c614367bebef0f2c41a}{vrna\+\_\+sc\+\_\+remove()}, \hyperlink{group__SHAPE__reactivities_gaf3c65a045060aef5c4e41693d30af58c}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam()}, \hyperlink{group__perturbation_gaa124bdc20d88001c38ade590c4bcc3c4}{vrna\+\_\+sc\+\_\+minimize\+\_\+pertubation()}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em vc} & The \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} the soft constraints are associated with \\
\hline
{\em reactivities} & A vector of normalized S\+H\+A\+P\+E reactivities \\
\hline
{\em m} & The slope of the conversion function \\
\hline
{\em b} & The intercept of the conversion function \\
\hline
{\em options} & The options flag indicating how/where to store the soft constraints \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
1 on successful extraction of the method, 0 on errors 
\end{DoxyReturn}
\hypertarget{group__SHAPE__reactivities_ga04ba85da63d8c793bb8001d1e6f800ba}{}\index{Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}!vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali@{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali}}
\index{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali@{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali}!Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}}
\subsubsection[{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali}]{\setlength{\rightskip}{0pt plus 5cm}int vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali (
\begin{DoxyParamCaption}
\item[{{\bf vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$}]{vc, }
\item[{const char $\ast$$\ast$}]{shape\+\_\+files, }
\item[{const int $\ast$}]{shape\+\_\+file\+\_\+association, }
\item[{double}]{m, }
\item[{double}]{b, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\label{group__SHAPE__reactivities_ga04ba85da63d8c793bb8001d1e6f800ba}


{\ttfamily \#include $<$\hyperlink{constraints__SHAPE_8h}{Vienna\+R\+N\+A/constraints\+\_\+\+S\+H\+A\+P\+E.\+h}$>$}



Add S\+H\+A\+P\+E reactivity data from files as soft constraints for consensus structure prediction (Deigan et al. method) 


\begin{DoxyParams}{Parameters}
{\em vc} & The \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} the soft constraints are associated with \\
\hline
{\em shape\+\_\+files} & A set of filenames that contain normalized S\+H\+A\+P\+E reactivity data \\
\hline
{\em shape\+\_\+file\+\_\+association} & An array of integers that associate the files with sequences in the alignment \\
\hline
{\em m} & The slope of the conversion function \\
\hline
{\em b} & The intercept of the conversion function \\
\hline
{\em options} & The options flag indicating how/where to store the soft constraints \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
1 on successful extraction of the method, 0 on errors 
\end{DoxyReturn}
\hypertarget{group__SHAPE__reactivities_gaf3c65a045060aef5c4e41693d30af58c}{}\index{Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}!vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam@{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam}}
\index{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam@{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam}!Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}}
\subsubsection[{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam}]{\setlength{\rightskip}{0pt plus 5cm}int vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam (
\begin{DoxyParamCaption}
\item[{{\bf vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$}]{vc, }
\item[{const double $\ast$}]{reactivities, }
\item[{double}]{b, }
\item[{double}]{default\+\_\+value, }
\item[{const char $\ast$}]{shape\+\_\+conversion, }
\item[{unsigned int}]{options}
\end{DoxyParamCaption}
)}\label{group__SHAPE__reactivities_gaf3c65a045060aef5c4e41693d30af58c}


{\ttfamily \#include $<$\hyperlink{constraints__SHAPE_8h}{Vienna\+R\+N\+A/constraints\+\_\+\+S\+H\+A\+P\+E.\+h}$>$}



Add S\+H\+A\+P\+E reactivity data as soft constraints (Zarringhalam et al. method) 

This method first converts the observed S\+H\+A\+P\+E reactivity of nucleotide $ i $ into a probability $ q_i $ that position $ i $ is unpaired by means of a non-\/linear map. Then pseudo-\/energies of the form \[ \Delta G_{\text{SHAPE}}(x,i) = \beta\ |x_i - q_i| \] are computed, where $ x_i=0 $ if position $ i $ is unpaired and $ x_i=1 $ if $ i $ is paired in a given secondary structure. The parameter $ \beta $ serves as scaling factor. The magnitude of discrepancy between prediction and experimental observation is represented by $ |x_i - q_i| $.

\begin{DoxySeeAlso}{See also}
For further details, we refer to \cite{zarringhalam:2012} 

\hyperlink{group__soft__constraints_ga73cdc07b9a199c614367bebef0f2c41a}{vrna\+\_\+sc\+\_\+remove()}, \hyperlink{group__SHAPE__reactivities_ga57d612b58e1c61dd6cfcb5a843f8f1b3}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan()}, \hyperlink{group__perturbation_gaa124bdc20d88001c38ade590c4bcc3c4}{vrna\+\_\+sc\+\_\+minimize\+\_\+pertubation()}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em vc} & The \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} the soft constraints are associated with \\
\hline
{\em reactivities} & A vector of normalized S\+H\+A\+P\+E reactivities \\
\hline
{\em b} & The scaling factor $ \beta $ of the conversion function \\
\hline
{\em options} & The options flag indicating how/where to store the soft constraints \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
1 on successful extraction of the method, 0 on errors 
\end{DoxyReturn}
\hypertarget{group__SHAPE__reactivities_ga67675b3ed48744489a3bcfa4174197cb}{}\index{Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}!vrna\+\_\+sc\+\_\+\+S\+H\+A\+P\+E\+\_\+to\+\_\+pr@{vrna\+\_\+sc\+\_\+\+S\+H\+A\+P\+E\+\_\+to\+\_\+pr}}
\index{vrna\+\_\+sc\+\_\+\+S\+H\+A\+P\+E\+\_\+to\+\_\+pr@{vrna\+\_\+sc\+\_\+\+S\+H\+A\+P\+E\+\_\+to\+\_\+pr}!Incorporating S\+H\+A\+P\+E reactivity data@{Incorporating S\+H\+A\+P\+E reactivity data}}
\subsubsection[{vrna\+\_\+sc\+\_\+\+S\+H\+A\+P\+E\+\_\+to\+\_\+pr}]{\setlength{\rightskip}{0pt plus 5cm}int vrna\+\_\+sc\+\_\+\+S\+H\+A\+P\+E\+\_\+to\+\_\+pr (
\begin{DoxyParamCaption}
\item[{const char $\ast$}]{shape\+\_\+conversion, }
\item[{double $\ast$}]{values, }
\item[{int}]{length, }
\item[{double}]{default\+\_\+value}
\end{DoxyParamCaption}
)}\label{group__SHAPE__reactivities_ga67675b3ed48744489a3bcfa4174197cb}


{\ttfamily \#include $<$\hyperlink{constraints__SHAPE_8h}{Vienna\+R\+N\+A/constraints\+\_\+\+S\+H\+A\+P\+E.\+h}$>$}



Convert S\+H\+A\+P\+E reactivity values to probabilities for being unpaired. 

This function parses the informations from a given file and stores the result in the preallocated string sequence and the \hyperlink{group__data__structures_ga31125aeace516926bf7f251f759b6126}{F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+B\+L} array values.

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga646ebf45450a69a7f2533f9ecd283a32}{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read()} 
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em shape\+\_\+conversion} & String definining the method used for the conversion process \\
\hline
{\em values} & Pointer to an array of S\+H\+A\+P\+E reactivities \\
\hline
{\em length} & Length of the array of S\+H\+A\+P\+E reactivities \\
\hline
{\em default\+\_\+value} & Result used for position with invalid/missing reactivity values \\
\hline
\end{DoxyParams}
